base/floatfuncs.jl

```diff
@@ -348,19 +351,14 @@ function fma_emulated(a::Float64, b::Float64,c::Float64)
     s = (abs(abhi) > abs(c)) ? (abhi-r+c+ablo) : (c-r+abhi+ablo)
     return r+s
 end
-fma_llvm(x::Float32, y::Float32, z::Float32) = fma_float(x, y, z)
-fma_llvm(x::Float64, y::Float64, z::Float64) = fma_float(x, y, z)
 
 # Disable LLVM's fma if it is incorrect, e.g. because LLVM falls back
 # onto a broken system libm; if so, use a software emulated fma
-@assume_effects :consistent fma(x::Float32, y::Float32, z::Float32) = Core.Intrinsics.have_fma(Float32) ? fma_llvm(x,y,z) : fma_emulated(x,y,z)
-@assume_effects :consistent fma(x::Float64, y::Float64, z::Float64) = Core.Intrinsics.have_fma(Float64) ? fma_llvm(x,y,z) : fma_emulated(x,y,z)
-
-function fma(a::Float16, b::Float16, c::Float16)
-    Float16(muladd(Float32(a), Float32(b), Float32(c))) #don't use fma if the hardware doesn't have it.
+@assume_effects :consistent function fma(x::T, y::T, z::T) where {T<:IEEEFloat}
+    Core.Intrinsics.have_fma(T) ? fma_float(x,y,z) : fma_emulated(x,y,z)
```

Comment:
My understanding is that `Core.Intrinsics.have_fma(Float16)` is always false at the moment because https://github.com/JuliaLang/julia/blob/9b1ea1a880e1c47ebdc549a12fca288b5cc60013/src/llvm-cpufeatures.cpp#L50-L76 only checks Float32/Float64 extensions. @gbaraldi is that correct?

Reply:
Actually, this may work on riscv64 with #57043, but I'm still not entirely sure about what's going on there.

Reply:
sure, at which point this is nfc for such architectures, but that can be fixed in a separate pr.